VectorBuilder#addAll and VectorIterator#copyToArray/addAll use System.arraycopy

joshlemer opened the pull request:
This optimizes VectorBuilder#addAll and VectorIterator#copyToArray to use System.arraycopy rather than copying one-by-one, resulting in a pretty large improvement, if my benchmarks are valid:

```scala
  @Param(Array("0", "10", "1000", "1000000"))
  var size: Int = _
  var first: Vector[Int] = _
  var second: Vector[Int] = _

  @Setup(Level.Trial) def initKeys(): Unit = {
    first = (0 until size).toVector
    second = (0 until size).toVector
  }

  @Benchmark def concat(bh: Blackhole): Any = {
    bh.consume(first ++ second)
  }
```

```
With Optimization:

[info] Benchmark                (size)  Mode  Cnt         Score        Error  Units

[info] VectorBenchmark.concat        0  avgt   20       48.523 ±      1.596  ns/op
[info] VectorBenchmark.concat       10  avgt   20       65.017 ±      1.183  ns/op
[info] VectorBenchmark.concat     1000  avgt   20     2424.475 ±    108.181  ns/op
[info] VectorBenchmark.concat  1000000  avgt   20  6774195.422 ± 147708.690  ns/op


2.13.x

[info] Benchmark                (size)  Mode  Cnt         Score         Error  Units

[info] Benchmark                (size)  Mode  Cnt        Score        Error  Units
[info] VectorBenchmark.concat        0  avgt   20       51.239 ±      5.362  ns/op
[info] VectorBenchmark.concat       10  avgt   20      102.522 ±     23.212  ns/op
[info] VectorBenchmark.concat     1000  avgt   20     5844.086 ±    435.703  ns/op
[info] VectorBenchmark.concat  1000000  avgt   20  9860384.444 ± 165811.511  ns/op
```

mkeskells commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -580,6 +580,31 @@ class VectorIterator[+A](_startIndex: Int, endIndex: Int)
 
     res
   }
+  override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Int = {
```
as display is not specialised will this work for copy to primitive types?

I think that Vector[Int] has display0 as Array[AnyRef], with casting & boxing

I think you will need to special case the AnyRef with the code below, and do super.copyToArray for the rest
@retronym is there a utility to copy with scala casting if the array types don't match?

joshlemer replied:
@mkeskells It does in fact work at least for `Int`, since that's what the benchmark was using

mkeskells replied:
@joshlemer this is not doing `copyToArray(Array[Int] ….)`
you are letting scala box and unbox.

```
  @Benchmark def concat(bh: Blackhole): Any = {
    bh.consume(first ++ second)
  }

```

in VectorPointer
`    private[immutable] var display0: Array[AnyRef] = _ `

joshlemer replied:
Aren't the elements boxed already because they are in Vector's `Array[AnyRef]`?

retronym replied:
Yes, `Vector[Int]` contains boxes.

mkeskells replied:
but to parameter to copyToArray could be a primitive array.

mkeskells commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -580,6 +580,31 @@ class VectorIterator[+A](_startIndex: Int, endIndex: Int)
+          gotoNextBlockStart(newBlockIndex, blockIndex ^ newBlockIndex)
+
+          blockIndex = newBlockIndex
+          endLo = math.min(endIndex - blockIndex, 32)
```
Math.min is faster

mkeskells commented:
@joshlemer would the same improvement work for the Vector.copyToArray. I don't think there is specialised code for that either

joshlemer commented:
@mkeskells yes a similar implementation could be done on Vector, i'll add that to this PR :-)

mkeskells commented:
I thing the benchmark would be better with Reference types, and using them should only make the comparison more pronounced if there is boxing going on (don't think that there is though)

I think a benchmark of the copyToArray would be more pronounced ( without the overheads of adding to another Vector)

retronym commented on src/library/scala/collection/immutable/Vector.scala:
```diff
@@ -553,33 +555,50 @@ class VectorIterator[+A](_startIndex: Int, endIndex: Int)
+    var totalCopied = 0
+    var _start = start
+    while (hasNext && totalCopied < len) {
+      val toBeCopied = (len - totalCopied) min (endLo - lo)
```
Might be able to use `IterableOnce.elemsToCopyToArray` here.

joshlemer replied:
Indeed, this was the source of the bug :-)

joshlemer commented:
@SethTisue we could add the following tags: Performance, Collections, Library

joshlemer commented:
Update: No longer is there an `IndexOutOfBounds` bug

joshlemer committed:
VectorBuilder and Vector copyToArray/addAll use System.arraycopy

Small refactor in VectorBuilder

Add VectorBenchmark

Use java.lang.Math.min instead of scala.math.min in VectorIterator

Factor out VectorIterator#advanceToNextBlockIfNecessary

Fix array index out of bounds exception in Vector#copyToArray

VectorBuilder and Vector copyToArray/addAll use System.arraycopy

Small refactor in VectorBuilder

Add VectorBenchmark

Use java.lang.Math.min instead of scala.math.min in VectorIterator

Factor out VectorIterator#advanceToNextBlockIfNecessary

Fix array index out of bounds exception in Vector#copyToArray

Fix array index out of bounds error in Vector#copyToArray

Small refactoring in Vector#copyToArray

VectorIterator#copyToArray uses Array.copy

remove git merge artifacts

Simplify VectorBuilder#addAll

Remove commented code in VectorBenchmark

joshlemer commented:
@szeiger is it alright if I merge this?
